### PR TITLE
Qinchen/Add GPU label support basing on the FM timeout implementation

### PR DESCRIPTION
**Pull request by qinchen123**

### CODE REVIEW

**Comment by qinchen123** [comment]
@fanyangCS is added to the review. #Closed

**Comment by qinchen123** [comment]
@abuccts is added to the review. #Closed

**Comment by yqwang-ms** [comment]
>       candidateNode.allocateResource(request, candidateSelectGPU);

1. rename to addContainerRequest
2. Should be called between:
    rmClient.addContainerRequest(request);
+ -> node.addContainerRequest()
    statusManager.transitionTaskState(taskLocator, TaskState.CONTAINER_REQUESTED,
        new TaskEvent().setContainerRequest(request));


---
Refers to: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java:114 in c41f626. [](commit_id = c41f6262f6b02247a99e6b2c50ab650887d58653, deletion_comment = False)

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/Node.java [hunk 1]
```diff
@@ -25,9 +25,8 @@
   private ResourceDescriptor capacity;
   private final String name;
   private ResourceDescriptor used;
-  //localAllocated in remember the local tried request, it will remember the tried information don't re-try the same request in a AM life cycle.
-  //so here, it is no need to de-crease the localAllocated information once the container request success.
-  private ResourceDescriptor localAllocated;
+  //localTriedAllocateResource: remember the local tried request, it will remember the tried information don't re-try the same request in a AM life cycle.
+  private ResourceDescriptor localAllocatingResource;
   private long selectedGpuBitmap;
   private Set<String> nodeLabels;
```
Comment (yqwang-ms) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/Node.java]: Remove

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/Node.java [hunk 2]
```diff
@@ -25,9 +25,8 @@
   private ResourceDescriptor capacity;
   private final String name;
   private ResourceDescriptor used;
-  //localAllocated in remember the local tried request, it will remember the tried information don't re-try the same request in a AM life cycle.
-  //so here, it is no need to de-crease the localAllocated information once the container request success.
-  private ResourceDescriptor localAllocated;
+  //localTriedAllocateResource: remember the local tried request, it will remember the tried information don't re-try the same request in a AM life cycle.
+  private ResourceDescriptor localAllocatingResource;
```
Comment (yqwang-ms) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/Node.java]: >localAllocatingResource [](start = 29, length = 23)

rename to requested

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/Node.java [hunk 3]
```diff
@@ -66,40 +65,47 @@ public int getTotalNumGpus() {
...
-    localAllocated.setGpuAttribute(localAllocated.getGpuAttribute() | gpuMap);
-    localAllocated.setGpuNumber(localAllocated.getGpuNumber() + resource.getGpuNumber());
+    localAllocatingResource.setCpuNumber(localAllocatingResource.getCpuNumber() + resource.getCpuNumber());
+    localAllocatingResource.setMemoryMB(localAllocatingResource.getMemoryMB() + resource.getMemoryMB());
+    localAllocatingResource.setGpuAttribute(localAllocatingResource.getGpuAttribute() | gpuMap);
+    localAllocatingResource.setGpuNumber(localAllocatingResource.getGpuNumber() + resource.getGpuNumber());
     selectedGpuBitmap = gpuMap;
   }
 
+  public void releaseLocalAllocatingResource(ResourceDescriptor resource) {
```
Comment (yqwang-ms) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/Node.java]: >releaseLocalAllocatingResource [](start = 14, length = 30)

rename to removeContainerRequest

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/Node.java [hunk 4]
```diff
@@ -25,9 +25,8 @@
   private ResourceDescriptor capacity;
   private final String name;
```
Comment (yqwang-ms) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/Node.java]: >name [](start = 23, length = 4)

hostName

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java [hunk 5]
```diff
@@ -986,13 +986,18 @@ private void removeContainerRequest(TaskStatus taskStatus) {
...
     if (request == null) {
       return;
     }
 
     try {
+
+      ResourceDescriptor resourceDescriptor = ResourceDescriptor.fromResource(request.getCapability());
+      for (String node : request.getNodes()) {
+        gpuAllocationManager.releaseLocalAllocatingResource(resourceDescriptor, node);
+      }
```
Comment (yqwang-ms) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java]: Move to another try catch

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java [hunk 6]
```diff
@@ -1,32 +1,36 @@
...
 // The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 //
-// THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING 
-// BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND 
-// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, 
-// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
-// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE. 
+// THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+// BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
```
Comment (yqwang-ms) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java]: No formating

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java [hunk 7]
```diff
@@ -73,14 +77,28 @@ public synchronized Node allocateCandidateRequestNode(ResourceDescriptor request
 
       if (nodeLabel != null) {
         Set<String> nodeLabels = entry.getValue().getNodeLabels();
-        if (!nodeLabels.contains(nodeLabel)) {
-          LOGGER.logInfo(
+        if (nodeLabels != null && nodeLabels.size() > 0 && !nodeLabels.contains(nodeLabel)) {
+          LOGGER.logDebug(
               "allocateCandidateRequestNode: Skip node %s, label does not match:%s",
               entry.getValue().getHostName(), nodeLabel);
```
Comment (yqwang-ms) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java]: log nodeLabels

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java [hunk 8]
```diff
@@ -73,14 +77,28 @@ public synchronized Node allocateCandidateRequestNode(ResourceDescriptor request
...
+        if (!nodes.containsKey(entry.getValue().getHostName())) {
+          LOGGER.logWarning(
+              "allocateCandidateRequestNode: Skip node %s, getGpuType not set",
+              entry.getValue().getHostName());
+          continue;
+        }
+        String gpuType = (nodes.get(entry.getValue().getHostName())).getGpuType();
+        if (!nodeGpuType.equals(gpuType)) {
+          LOGGER.logDebug(
+              "allocateCandidateRequestNode: Skip node %s (gpuTypeLabel:%s), labels don't match: request nodeLabel: %s",
```
Comment (yqwang-ms) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java]: >(gpuTypeLabel:%s), labels don't match: request nodeLabel: %s", [](start = 58, length = 62)

gpuType

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/Node.java (hunk 2 above)
Comment (yqwang-ms) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/Node.java]: remove

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/Node.java [hunk 9]
```diff
@@ -66,40 +65,47 @@ public int getTotalNumGpus() {
...
   public int getAvailableCpu() {
-    return capacity.getCpuNumber() - used.getCpuNumber() - localAllocated.getCpuNumber();
+    return capacity.getCpuNumber() - used.getCpuNumber() - localAllocatingResource.getCpuNumber();
   }
 
   public long getSelectedGpuBitmap() {
     return selectedGpuBitmap;
   }
 
   public void allocateResource(ResourceDescriptor resource, long gpuMap) {
```
Comment (yqwang-ms) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/Node.java]: >, long gpuMap [](start = 58, length = 13)

remove

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java (hunk 6 above)
Comment (qinchen123) [reply frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java]: Previous version which space is not standard. 

---
In reply to: [158486837](https://github.com/Microsoft/pai/pull/128#discussion_r158486837) [](ancestors = 158486837)

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/SelectionResult.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+//
+// THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+// BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+package com.microsoft.frameworklauncher.applicationmaster;
+
+public class SelectionResult {
```
Comment (yqwang-ms) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/SelectionResult.java]: >SelectionResult [](start = 13, length = 15)

rename to GpuAllocation

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java [hunk 11]
```diff
@@ -336,10 +336,10 @@ private ContainerRequest setupContainerRequest(TaskStatus taskStatus) throws Exc
     if (resource.getGpuNumber() > 0 && resource.getGpuAttribute() == 0) {
       updateNodeReport(yarnClient.getNodeReports(NodeState.RUNNING));
 
-      Node node = gpuAllocationManager.allocateCandidateRequestNode(resource, nodeLabel, nodeGpuType);
-      if (node != null) {
-        resource.setGpuAttribute(node.getSelectedGpuBitmap());
-        return HadoopUtils.toContainerRequest(resource, priority, null, node.getHostName());
+      SelectionResult selectionResult = gpuAllocationManager.SelectCandidateRequestNode(resource, nodeLabel, nodeGpuType);
```
Comment (yqwang-ms) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java]: >SelectCandidateRequestNode [](start = 61, length = 26)

low case the method

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java [hunk 12]
```diff
@@ -336,10 +336,10 @@ private ContainerRequest setupContainerRequest(TaskStatus taskStatus) throws Exc
...
+      if (selectionResult != null) {
+        resource.setGpuAttribute(selectionResult.getSelectedGpuBitmap());
```
Comment (yqwang-ms) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java]: >getSelectedGpuBitmap [](start = 49, length = 20)

getGpuBitmap

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java [hunk 13]
```diff
@@ -997,6 +1000,14 @@ private void removeContainerRequest(TaskStatus taskStatus) {
     } catch (Exception e) {
       LOGGER.logError(e, "%s: Failed to removeContainerRequest", taskLocator);
     }
+    if(request.getNodes() != null && request.getNodes().size() != 0) {
+      try {
+        ResourceDescriptor resourceDescriptor = ResourceDescriptor.fromResource(request.getCapability());
```
Comment (yqwang-ms) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java]: move before removeContainerRequest

**Comment by yqwang-ms** [comment]
>       candidateNode.allocateResource(request, candidateSelectGPU);

take care the position

---
In reply to: [353563872](https://github.com/Microsoft/pai/pull/128#issuecomment-353563872) [](ancestors = 353563872)

---
Refers to: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java:114 in c41f626. [](commit_id = c41f6262f6b02247a99e6b2c50ab650887d58653, deletion_comment = False)

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/Node.java (hunk 3 above)
Comment (yqwang-ms) [reply frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/Node.java]: align with others

---
In reply to: [158463022](https://github.com/Microsoft/pai/pull/128#discussion_r158463022) [](ancestors = 158463022)

**Comment by yqwang-ms** [comment]
>       candidateNode.allocateResource(request, candidateSelectGPU);

1. rename to addContainerRequest

---
In reply to: [353563872](https://github.com/Microsoft/pai/pull/128#issuecomment-353563872) [](ancestors = 353563872)

---
Refers to: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java:114 in c41f626. [](commit_id = c41f6262f6b02247a99e6b2c50ab650887d58653, deletion_comment = False)

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java (hunk 6 above)
Comment (yqwang-ms) [reply frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java]: pls align with others

---
In reply to: [158578189](https://github.com/Microsoft/pai/pull/128#discussion_r158578189) [](ancestors = 158578189,158486837)

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java (hunk 7 above)
Comment (yqwang-ms) [reply frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java]: bugs

---
In reply to: [158486868](https://github.com/Microsoft/pai/pull/128#discussion_r158486868) [](ancestors = 158486868)

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java (hunk 8 above)
Comment (yqwang-ms) [reply frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java]: remove label pls

---
In reply to: [158486901](https://github.com/Microsoft/pai/pull/128#discussion_r158486901) [](ancestors = 158486901)

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java [hunk 14]
```diff
@@ -997,6 +1000,14 @@ private void removeContainerRequest(TaskStatus taskStatus) {
...
+        gpuAllocationManager.releaseRequestingResource(resourceDescriptor, request.getNodes());
+      } catch (Exception e) {
+        LOGGER.logError(e, "%s: Failed to ResourceDescriptor.fromResource", taskLocator);
```
Comment (yqwang-ms) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java]: >ResourceDescriptor.fromResource [](start = 42, length = 31)

can still consider it as a whole, so:
LOGGER.logError(e, "%s: Failed to releaseRequestingResource

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/Node.java [hunk 15]
```diff
@@ -54,56 +52,66 @@ public ResourceDescriptor getUsedResource() {
...
+    requestingResource.setGpuNumber(requestingResource.getGpuNumber() - resource.getGpuNumber());
   }
 
-  public void allocateResource(ResourceDescriptor resource, long gpuMap) {
-    localAllocated.setCpuNumber(localAllocated.getCpuNumber() + resource.getCpuNumber());
-    localAllocated.setMemoryMB(localAllocated.getMemoryMB() + resource.getMemoryMB());
-    localAllocated.setGpuAttribute(localAllocated.getGpuAttribute() | gpuMap);
-    localAllocated.setGpuNumber(localAllocated.getGpuNumber() + resource.getGpuNumber());
-    selectedGpuBitmap = gpuMap;
+  public void releaseLocalAllocatingResource(ResourceDescriptor resource) {
```
Comment (qinchen123) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/Node.java]: >releaseLocalAllocatingResource [](start = 14, length = 30)

Duplicated

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/SelectionResult.java [hunk 16]
```diff
@@ -0,0 +1,39 @@
+// Copyright (c) Microsoft Corporation
+// All rights reserved.
+//
+// MIT License
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
```
Comment (qinchen123) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/SelectionResult.java]: copy from other file

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java [hunk 17]
```diff
@@ -51,70 +57,89 @@ public synchronized void addCandidateRequestNode(Node candidateRequestNode) {
...
+          if (!nodes.containsKey(entry.getValue().getHostName())) {
+            LOGGER.logWarning(
+                "SelectCandidateRequestNode: Skip node %s, getGpuType not set",
+                entry.getValue().getHostName());
+            continue;
+          }
+          String gpuType = (nodes.get(entry.getValue().getHostName())).getGpuType();
+          if (!nodeGpuType.equals(gpuType)) {
+            LOGGER.logDebug(
+                "SelectCandidateRequestNode: Skip node %s (gpuTypeLabel:%s), gpuType don't match: request gpuType: %s",
```
Comment (qinchen123) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java]: >gpuTypeLabel [](start = 59, length = 12)

gpuType

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java [hunk 18]
```diff
@@ -683,6 +683,10 @@ private void addContainerRequest(TaskStatus taskStatus) throws Exception {
     statusManager.transitionTaskState(taskLocator, TaskState.CONTAINER_REQUESTED,
         new TaskEvent().setContainerRequest(request));
 
+    if(request.getNodes().size() > 0) {
+      gpuAllocationManager.allocateRequestingResource(ResourceDescriptor.fromResource(request.getCapability()), request.getNodes());
+    }
```
Comment (qinchen123) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java]: move up

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java [hunk 19]
```diff
@@ -683,6 +683,10 @@ private void addContainerRequest(TaskStatus taskStatus) throws Exception {
     statusManager.transitionTaskState(taskLocator, TaskState.CONTAINER_REQUESTED,
```
Comment (qinchen123) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java]: here

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java [hunk 20]
```diff
@@ -683,6 +683,10 @@ private void addContainerRequest(TaskStatus taskStatus) throws Exception {
...
         new TaskEvent().setContainerRequest(request));
 
+    if(request.getNodes().size() > 0) {
+      gpuAllocationManager.allocateRequestingResource(ResourceDescriptor.fromResource(request.getCapability()), request.getNodes());
```
Comment (qinchen123) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java]: >allocateRequestingResource [](start = 27, length = 26)

rename to addContainerRequest

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java [hunk 21]
```diff
@@ -680,6 +680,9 @@ private void addContainerRequest(TaskStatus taskStatus) throws Exception {
     LOGGER.logInfo("%s: addContainerRequest with timeout %ss. ContainerRequest: [%s]",
         taskLocator, containerRequestTimeoutSec, HadoopExtensions.toString(request));
     rmClient.addContainerRequest(request);
+    if(request.getNodes().size() > 0) {
+      gpuAllocationManager.addContainerRequest(ResourceDescriptor.fromResource(request.getCapability()), request.getNodes());
+    }
```
Comment (yqwang-ms) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java]: gpuAllocationManager.addContainerRequest(request);

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java [hunk 22]
```diff
@@ -997,6 +999,14 @@ private void removeContainerRequest(TaskStatus taskStatus) {
...
       LOGGER.logError(e, "%s: Failed to removeContainerRequest", taskLocator);
     }
+    if(request.getNodes() != null && request.getNodes().size() != 0) {
+      try {
+        ResourceDescriptor resourceDescriptor = ResourceDescriptor.fromResource(request.getCapability());
+        gpuAllocationManager.removeContainerRequest(resourceDescriptor, request.getNodes());
+      } catch (Exception e) {
+        LOGGER.logError(e, "%s: Failed to gpuAllocationManager.removeContainerRequest", taskLocator);
+      }
+    }
```
Comment (yqwang-ms) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java]: try {
      rmClient.removeContainerRequest(request);
    } catch (Exception e) {
      LOGGER.logError(e, "%s: Failed to rmClient.removeContainerRequest", taskLocator);
    }

    try {
      gpuAllocationManager.removeContainerRequest(request);
    } catch (Exception e) {
      LOGGER.logError(e, "%s: Failed to gpuAllocationManager.removeContainerRequest", taskLocator);
    }

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java [hunk 23]
```diff
@@ -51,70 +58,89 @@ public synchronized void addCandidateRequestNode(Node candidateRequestNode) {
...
 
       if (nodeLabel != null) {
         Set<String> nodeLabels = entry.getValue().getNodeLabels();
-        if (!nodeLabels.contains(nodeLabel)) {
-          LOGGER.logInfo(
-              "allocateCandidateRequestNode: Skip node %s, label does not match:%s",
-              entry.getValue().getHostName(), nodeLabel);
+        if (nodeLabels != null && nodeLabels.size() > 0 && !nodeLabels.contains(nodeLabel)) {
+          LOGGER.logDebug(
+              "SelectCandidateRequestNode: Skip node %s (nodeLabels: %s), label does not match: request label: %s",
```
Comment (yqwang-ms) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java]: >SelectCandidateRequestNode [](start = 15, length = 26)

lower case

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocation.java [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+
+  public void setNodeName(String nodeName) {
+    this.nodeName = nodeName;
+  }
+
+  public long getGpuBitmap() {
+    return gpuBitmap;
+  }
+
+  public void setSelectedGpuMap(Long gpuBitmap) {
```
Comment (yqwang-ms) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocation.java]: >setSelectedGpuMap [](start = 14, length = 17)

rename

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java [hunk 25]
```diff
@@ -51,70 +58,89 @@ public synchronized void addCandidateRequestNode(Node candidateRequestNode) {
...
```
Comment (yqwang-ms) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java]: >SelectCandidateRequestNode [](start = 15, length = 26)

lower case all places

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java (hunk 22 above)
Comment (qinchen123) [reply frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java]: What do you mean ?

---
In reply to: [158595991](https://github.com/Microsoft/pai/pull/128#discussion_r158595991) [](ancestors = 158595991)

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java [hunk 26]
```diff
@@ -997,6 +998,13 @@ private void removeContainerRequest(TaskStatus taskStatus) {
     } catch (Exception e) {
       LOGGER.logError(e, "%s: Failed to removeContainerRequest", taskLocator);
```
Comment (yqwang-ms) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java]: >removeContainerRequest [](start = 40, length = 22)

rmClient.removeContainerRequest

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java [hunk 27]
```diff
@@ -336,10 +336,10 @@ private ContainerRequest setupContainerRequest(TaskStatus taskStatus) throws Exc
     if (resource.getGpuNumber() > 0 && resource.getGpuAttribute() == 0) {
       updateNodeReport(yarnClient.getNodeReports(NodeState.RUNNING));
 
-      Node node = gpuAllocationManager.allocateCandidateRequestNode(resource, nodeLabel, nodeGpuType);
-      if (node != null) {
-        resource.setGpuAttribute(node.getSelectedGpuBitmap());
-        return HadoopUtils.toContainerRequest(resource, priority, null, node.getHostName());
+      GpuAllocation result = gpuAllocationManager.selectCandidateRequestNode(resource, nodeLabel, nodeGpuType);
```
Comment (yqwang-ms) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java]: >result [](start = 20, length = 6)

gpuAllocation

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java [hunk 28]
```diff
@@ -51,70 +59,89 @@ public synchronized void addCandidateRequestNode(Node candidateRequestNode) {
...
+          if (!nodes.containsKey(entry.getValue().getHostName())) {
+            LOGGER.logWarning(
+                "selectCandidateRequestNode: Skip node %s, getGpuType not set",
+                entry.getValue().getHostName());
+            continue;
+          }
+          String gpuType = (nodes.get(entry.getValue().getHostName())).getGpuType();
+          if (!nodeGpuType.equals(gpuType)) {
+            LOGGER.logDebug(
+                "selectCandidateRequestNode: Skip node %s (gpuType:%s), gpuType don't match: request gpuType: %s",
```
Comment (yqwang-ms) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java]: space

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java [hunk 29]
```diff
@@ -124,4 +151,36 @@ public synchronized void removeCandidateRequestNode(Node candidateRequestNode) {
...
+
+  public synchronized void addContainerRequest(AMRMClient.ContainerRequest request) throws Exception {
+
+    ResourceDescriptor resource = ResourceDescriptor.fromResource(request.getCapability());
+    List<String> nodeList = request.getNodes();
+
+    for (String nodeName : nodeList) {
+      if (!candidateRequestNodes.containsKey(nodeName)) {
+        LOGGER.logWarning(
+            "ReleaseLocalAllocatingResource: node is not exist: " + nodeName);
```
Comment (yqwang-ms) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java]: >ReleaseLocalAllocatingResource [](start = 13, length = 30)

addContainerRequest

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java [hunk 30]
```diff
@@ -124,4 +151,36 @@ public synchronized void removeCandidateRequestNode(Node candidateRequestNode) {
...
+
+  public synchronized void removeContainerRequest(AMRMClient.ContainerRequest request) throws Exception {
+
+    ResourceDescriptor resource = ResourceDescriptor.fromResource(request.getCapability());
+    List<String> nodeList = request.getNodes();
+
+    for (String nodeName : nodeList) {
+      if (!candidateRequestNodes.containsKey(nodeName)) {
+        LOGGER.logWarning(
+            "ReleaseLocalAllocatingResource: node is not exist: " + nodeName);
```
Comment (yqwang-ms) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java]: >ReleaseLocalAllocatingResource [](start = 13, length = 30)

removeContainerRequest

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java (hunk 28 above)
Comment (qinchen123) [reply frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java]: space what ?

---
In reply to: [158621176](https://github.com/Microsoft/pai/pull/128#discussion_r158621176) [](ancestors = 158621176)

Comment (yqwang-ms) [reply frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java]: add space:
(gpuType:%s)
->
(gpuType: %s)

---
In reply to: [158623074](https://github.com/Microsoft/pai/pull/128#discussion_r158623074) [](ancestors = 158623074,158621176)

**Pull request merged**